### DISCUSSION TITLE
React Native node and native dependency updates

### DISCUSSION THREAD

**Discussion by cranberyxl** [renovatebot/renovate#9317] [Request Help] [answered]
Is there a way to have cocoapods and gradle update in the same PR as a react native node dependency?

**Comment by rarkins** [comment]
Use `packageRules` to match the packages you want and then add a `groupName`

**Reply by cranberyxl** [reply]
How do I go from `groupName` to updating the `Podfile.lock`? Can you point to an example?

**Reply by rarkins** [reply]
Please create a sample repo to demonstrate what you mean.

**Reply by cranberyxl** [reply]
I created a basic react-native app here: https://github.com/cranberyxl/RenovateReactNativeExample

I've intentionally added an older version of `react-native-video` to `package.json`. As part of that I also run `pod install` in the `ios` directory to set the `Podfile.lock` se this commit: https://github.com/cranberyxl/RenovateReactNativeExample/commit/74456bfb3ed2a3c263f8b13109d679be1bc08f6d

Renovate will eventually open an update to `package.json` for the node dep. As part of that the `Podfile.lock` will need to be updated. I've been doing this manually, but if there's a way for renovate to understand to update both npm and cocoapods, that would be great. Additionally, we will start making gradle updates.

**Reply by rarkins** [reply]
Thanks, I understand better what you mean now. Will answer in a new thread as grouping is not really the solution here.

**Comment by rarkins** [comment, marked as answer]
Based on the above thread, it seems that your `Podfile.lock` (and maybe also your Gradle?) depends on the contents of `package.json`/`yarn.lock` so needs refreshing if they change. Unfortunately we do not support this concept today - `Podfile.lock` is only updated if `Podfile` is updated.

 If you self-host, you could make use of [postUpgradeTasks](https://docs.renovatebot.com/configuration-options/#postupgradetasks) to run any custom command after an upgrade, including `pod` ones.

**Reply by cranberyxl** [reply]
Thanks for the answer. Is this somethign you would consider supporting for the growing world of React Native and Flutter type applications that span npm, cocoapods, and gradle?

**Reply by rarkins** [reply]
Yes, we definitely could try to make this automatic. If you can create a detailed feature request issue to get things started, that would help (ideally with an example of something outdated like you have here). BTW we don't really have an answer about how to run some podfile updates on non-Mac platforms though. The hosted app uses Linux and Docker containers.

**Reply by jarnove** [reply]
Hi @rarkins , @cranberyxl ,

I tried it with the postUpgradeTasks but without any results for the moment.

We're working on a react-native project and using Renovate as self-hosted but most of the time Renovate doesn't listen to the postUpgradeTasks. When they are executed anyway I don't see any changes, but that's impossible because most of the time the `Podfile.lock` must be updated as well.

My code looks like this:

**Renovate CI:**
`"allowedPostUpgradeCommands": [
    "bundle exec pod install --project-directory=ios"
  ],`

**Project:**
`"postUpgradeTasks": {
    "commands": [
      "bundle exec pod install --project-directory=ios"
    ],
    "fileFilters": ["ios/Podfile.lock"]
  },`

If I'm correct, the postUpgradeTasks are executed in the project itself? Or am I missing something?

**Reply by viceice** [reply]
@jarnove How do you run renovate?

**Reply by jarnove** [reply]
Via Azure Devops pipelines. I tried it in all different ways, but because the inconsistency (most of the time the postUpgradeTasks aren't called) it's had to debug. I also tried to execute a sh script only to update a license file, but also without any results. Any ideas?

**Reply by rarkins** [reply]
> If I'm correct, the postUpgradeTasks are executed in the project itself? Or am I missing something?

Yes, you're right. Are you using `renovate/renovate` full image and not `:slim` ?

**Reply by viceice** [reply]
@jarnove Do you set [trustLevel](https://docs.renovatebot.com/self-hosted-configuration/#trustlevel) to `high`, it's currently a requirement for [postUpgradeTasks](https://docs.renovatebot.com/configuration-options/#postupgradetasks)

**Reply by jarnove** [reply]
@rarkins , we have a project on Azure Devops (renovate-cli) with renovate as devDependency in our package.json. No idea about the full image or :slim.

@viceice , do I need this in both project config files (`config.json` & `renovate.json`)? For the moment it's only added in the `config.json` of Renovate.

**Reply by viceice** [reply]
`trustLevel` needs to be in bot config only, as it's a admin config option like `allowedPostUpgradeCommands`

Admin options can also passed as environment variables

**Reply by jarnove** [reply]
I already added the trustLevel to the bot config, so that's not the issue.

**Reply by viceice** [reply]
So then we need debug logs, so add `LOG_LEVEL=debug` to your env and share the sanitized logs

**Reply by jarnove** [reply]
For the moment my postUpgradeTasks looks like this:

```
"postUpgradeTasks": {
    "commands": [
      "sudo gem install bundler",
      "bundle",
      "bundle exec pod install --project-directory=ios",
      "./scripts/maintenance/generate-licenses.sh"
    ],
    "fileFilters": ["ios/Podfile.lock", "android/app/src/main/assets/oss-licenses.json"]
},
```

Looks good? It complains about the bundle command without sudo but when I add sudo, it complains that I don't need it.

Some logs:

```
WARN: artifactErrors (repository=-----, branch=renovate/react-navigation-monorepo)
    "artifactErrors": [
      {
        "lockFile": "package.json",
        "stderr": "Command failed: bundle\nFollowing files may not be writable, so sudo is needed:\n  /usr/local/bin\n  /var/lib/gems/2.7.0\n  /var/lib/gems/2.7.0/build_info\n  /var/lib/gems/2.7.0/cache\n  /var/lib/gems/2.7.0/doc\n  /var/lib/gems/2.7.0/extensions\n  /var/lib/gems/2.7.0/gems\n  /var/lib/gems/2.7.0/specifications\nGem::Ext::BuildError: ERROR: Failed to build gem native extension.\n\ncurrent directory:\n/tmp/bundler20210407-2452-e3craycurb-0.9.11/gems/curb-0.9.11/ext\n/usr/bin/ruby2.7 -I /usr/lib/ruby/2.7.0 -r ./siteconf20210407-2452-1gazqw.rb\nextconf.rb\nchecking for curl-config... no\nchecking for -lcurl... no\n*** extconf.rb failed ***\nCould not create Makefile due to some reason, probably lack of necessary\nlibraries and/or headers.  Check the mkmf.log file for more details.  You may\nneed configuration options.\n\nProvided configuration options:\n\t--with-opt-dir\n\t--without-opt-dir\n\t--with-opt-include\n\t--without-opt-include=${opt-dir}/include\n\t--with-opt-lib\n\t--without-opt-lib=${opt-dir}/lib\n\t--with-make-prog\n\t--without-make-prog\n\t--srcdir=.\n\t--curdir\n\t--ruby=/usr/bin/$(RUBY_BASE_NAME)2.7\n\t--with-curl-dir\n\t--without-curl-dir\n\t--with-curl-include\n\t--without-curl-include=${curl-dir}/include\n\t--with-curl-lib\n\t--without-curl-lib=${curl-dir}/lib\n\t--with-curllib\n\t--without-curllib\nextconf.rb:18:in `<main>':   Can't find libcurl or curl/curl.h (RuntimeError)\n\n  Try passing --with-curl-dir or --with-curl-lib and --with-curl-include\n  options to extconf.\n\nTo see why this extension failed to compile, please check the mkmf.log which can\nbe found here:\n\n/tmp/bundler20210407-2452-e3craycurb-0.9.11/extensions/x86_64-linux/2.7.0/curb-0.9.11/mkmf.log\n\nextconf failed, exit code 1\n\nGem files will remain installed in\n/tmp/bundler20210407-2452-e3craycurb-0.9.11/gems/curb-0.9.11 for inspection.\nResults logged to\n/tmp/bundler20210407-2452-e3craycurb-0.9.11/extensions/x86_64-linux/2.7.0/curb-0.9.11/gem_make.out\n\nAn error occurred while installing curb (0.9.11), and Bundler cannot continue.\nMake sure that `gem install curb -v '0.9.11' --source 'https://rubygems.org/'`\nsucceeds before bundling.\n\nIn Gemfile:\n  fastlane-plugin-badge was resolved to 1.4.0, which depends on\n    badge was resolved to 0.12.0, which depends on\n      curb\n"
      },
      {
        "lockFile": "package.json",
        "stderr": "Command failed: bundle exec pod install --project-directory=ios\n/var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/spec_set.rb:89:in `block in materialize': Could not find curb-0.9.11 in any of the sources (Bundler::GemNotFound)\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/spec_set.rb:82:in `map!'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/spec_set.rb:82:in `materialize'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/definition.rb:185:in `specs'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/definition.rb:255:in `specs_for'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/definition.rb:237:in `requested_specs'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/runtime.rb:91:in `block in definition_method'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/runtime.rb:20:in `setup'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler.rb:148:in `setup'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/setup.rb:20:in `block in <top (required)>'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/ui/shell.rb:136:in `with_level'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/ui/shell.rb:88:in `silence'\n\tfrom /var/lib/gems/2.7.0/gems/bundler-2.2.15/lib/bundler/setup.rb:20:in `<top (required)>'\n\tfrom /usr/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'\n\tfrom /usr/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'\n"
      },
      {
        "lockFile": "package.json",
        "stderr": "Command failed: ./scripts/maintenance/generate-licenses.sh\nnpm ERR! Error while executing:\nnpm ERR! /usr/bin/git ls-remote -h -t https://github.com/mwittig/license-checker.git\nnpm ERR! \nnpm ERR! fatal: unable to access 'https://github.com/mwittig/license-checker.git/': The requested URL returned error: 400\nnpm ERR! \nnpm ERR! exited with error code: 128\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /home/vsts/.npm/_logs/2021-04-07T09_03_25_193Z-debug.log\nInstall for [ 'npm-license-scraper@latest' ] failed with code 1\n"
      },

**Reply by viceice** [reply]
Check your logs, it's the `generate-licenses.sh`  command which fails

**Reply by jarnove** [reply]
Indeed, that command is also failing, locally it works, but the pods aren't updated either because that command is also failing.

**Reply by viceice** [reply]
Please read your log files,: `Can't find libcurl or curl/curl.h` it clearly says that some deps are missing, which need to be installed before running bundler.
